### PR TITLE
New discount module

### PR DESCRIPTION
**Pull request by chessbr**
Replacement for catalog campaigns. Catalog campaigns way too complex and adding few campaigns starts to take a toll on performance. There just doesn't seem to be any sane way to improve catalog campaigns either.

This should implement same discount options. Also few extras like discounted price and availability exceptions.

This has worse extendability but in theory the queryset in the module could be extendable by providing query filters. But still some weird discounts can be implemented as separate discount modules which I guess is the way it was supposed to be instead of making the default discount module too complex.

Goal for this discount module is to make product discounts lightning fast without losing the possibilities the catalog campaigns offer. Getting the discounts with one query should take care of the performance part.

For the future we likely want similar thing for basket campaigns. Also we should probably extend coupon codes into this one.

TODO:

* [x] Admin for happy hours
* [x] Admin for archives and delete
* [x] Admin for availability exceptions
* [x] Tests for admin views and forms
* [x] Improve catalog campaign sync. At least happy hours is not yet synced.
* [x] Add data migration for catalog campaign sync (nope, related fields are problematic with data migration so better just offer management command which also makes the sync optional)
* [x] Translations
* [x] Quick add for contact groups and filter groups with shop (waiting https://github.com/shuup/shuup/pull/1409)

Future ides:
* Add basket qty condition. Set product discounts based on the product qty in basket
* Combine basket campaigns into this. In theory we could link effect to the discount model. This effect would be exactly same effect we have already in basket campaigns. Then on product discount we can filter out discounts which have effect set. And for basket campaigns we can still likely loop the matching effects for best discount. As long as we get those effects with one query. Effects could be added to discount through quick add keeping the admin super simple.
* Maybe we should have select2 fields for things like categories, products, contacts and contact groups. Now we only have these select2multiple-fields. On the other hand we don't have quick add widget for these select2multiple-fields.
* Contact quick create

Questions:
* Should we try to simplify this even more and maybe just use shop instead of shops for discounts for example?

Refs https://github.com/shuup/shuup/issues/1163

Replaces https://github.com/shuup/shuup/pull/1405

### CODE REVIEW

File: shuup/discounts/utils.py [hunk 1]
```diff
@@ -57,17 +64,98 @@ def get_potential_discounts_for_product(context, product):
...
+        next_week_date = next_date + datetime.timedelta(days=7)
+        ranges.extend([
+            next_week_date.replace(hour=from_hour.hour, minute=from_hour.minute),
+            next_week_date.replace(hour=to_hour.hour, minute=to_hour.minute)
+        ])
+
+    return ranges
+
+
+def get_price_expiration(context, product):
```
Comment (tulimaki) [inline shuup/discounts/utils.py]: Problem! This doesn't look like being lighting fast operation. So why not return shared expiration date for the whole module based on active discounts. Now every time price for product expires we need to this "expensive" expire on calculation. This "shared" expiration date then is easy to cache and can be used for all prices coming through this discount module.

Likely the shop discounts are not timed the way that those would change in every 5min anyways.

File: shuup/core/utils/price_display.py [hunk 2]
```diff
@@ -82,30 +83,58 @@ def __call__(self, context, item, quantity=1, include_taxes=None, allow_cache=Tr
...
-        priceful = _get_priceful(context.get('request'), item, quantity)
-        if not priceful:
-            return ""
+        request = context.get('request')
+        price_info = get_cached_price_info(request, item, quantity) if allow_cache else None
+
+        if not price_info:
+            price_info = _get_priceful(request, item, quantity)
+
+            if price_info:
```
[the remark this answers is not in the document]
Comment (chessbr) [reply shuup/core/utils/price_display.py]: I guess no.. I want to return `""` just when there is no `price_info`, if `allow_cache` is True or False it doesn't matter.. :thinking:

[the remark this answers is not in the document]
Comment (chessbr) [reply shuup/core/utils/price_display.py]: but ok, I improved the `ifs` a little

File: shuup/discounts/modules.py [hunk 3]
```diff
@@ -59,20 +62,21 @@ def discount_price(self, context, product, price_info):
...
+            minimum_price_value = minimum_price_values[0] if minimum_price_values else 0
 
             price_info.price = max(
                 min(discounted_prices),
-                shop.create_price(minimum_price_value) or shop.create_price(0)
+                shop.create_price(minimum_price_value or 0) or shop.create_price(0)
             )
 
+        price_expiration = get_price_expiration(context, product)
+        if price_expiration and (not price_info.expires_on or price_expiration < price_info.expires_on):
```
[the remark this answers is not in the document]
Comment (chessbr) [reply shuup/discounts/modules.py]: isn't this what it is being doing in this line?

**Comment by tulimaki** [comment]
Tests seem to be unreliable. Maybe something to do with timezones.

File: shuup/discounts/utils.py [hunk 4]
```diff
@@ -57,17 +71,145 @@ def get_potential_discounts_for_product(context, product):
...
+    context_cache_key = "price_expiration_%(shop_id)s" % dict(shop_id=context.shop.pk)
+    if hasattr(context, "context_cache_key"):
+        return getattr(context, context_cache_key)
+
+    from shuup.discounts.models import AvailabilityException, Discount, TimeRange
+
+    if settings.SHUUP_DISCOUNTS_PER_PRODUCT_EXPIRATION_DATES:
+        potential_discounts = get_potential_discounts_for_product(context, product, available_only=False)
+    else:
+        potential_discounts = Discount.objects.filter(shops=context.shop, active=True)
```
Comment (tulimaki) [inline shuup/discounts/utils.py]: Should this also check wether the `end_datetime` is None or in future? We don't want to include discounts which dates is in past.

**Pull request merged**